File: quick/quick.jl
function _partition!(arr::Vector, low::Int, high::Int)
    pivot = arr[high]
    i = low - 1
    for j in low:(high - 1)
        if arr[j] < pivot
            i += 1
            arr[i], arr[j] = arr[j], arr[i]
        end
    end
    arr[i + 1], arr[high] = arr[high], arr[i + 1]
    return i + 1
end

function quicksort!(arr::Vector, low::Int = 1, high::Int = length(arr))
    if low < high
        pivot_index = _partition!(arr, low, high)
        quicksort!(arr, low, pivot_index - 1)
        quicksort!(arr, pivot_index + 1, high)
    end
    return arr
end

File: quick/quick.cpp
#include <iostream>
#include <vector>
#include <utility>

int partition(std::vector<int>& arr, int low, int high) {
    int pivot = arr[high];
    int i = low - 1;
    for (int j = low; j < high; ++j) {
        if (arr[j] <= pivot) {
            ++i;
            std::swap(arr[i], arr[j]);
        }
    }
    std::swap(arr[i + 1], arr[high]);
    return i + 1;
}

void quicksort(std::vector<int>& arr, int low, int high) {
    if (low < high) {
        int pi = partition(arr, low, high);
        quicksort(arr, low, pi - 1);
        quicksort(arr, pi + 1, high);
    }
}

void print_array(const std::vector<int>& arr) {
    for (const auto& v : arr) std::cout << v << ' ';
    std::cout << '\n';
}

int main() {
    std::vector<int> arr{10, 7, 8, 9, 1, 5};
    std::cout << "Original array: ";
    print_array(arr);
    quicksort(arr, 0, static_cast<int>(arr.size() - 1));
    std::cout << "Sorted array: ";
    print_array(arr);
    return 0;
}


